### PR TITLE
add usb redirdev device test case

### PR DESCRIPTION
**Pull request by brucepatricklee**
- by default, cold-plug/hot-unplug usb redirdev device to/from guest
 - add usb redirdev device by type spicevm or tcp
 - hot-plug/unplug usb redirdev device to/from guest

Signed-off-by: Jin Li <jil@redhat.com>

### CODE REVIEW

**Comment by brucepatricklee** [comment]
this commit should be merged with https://github.com/avocado-framework/avocado-vt/pull/2076

**Comment by qiankehan** [comment]
@brucepatricklee 
What is the cases' name ? I cannot find them:
```
avocado list --vt-type libvirt |grep spicevmc 
VT           type_specific.io-github-autotest-libvirt.serial.functional.type_spicevmc
```

**Comment by brucepatricklee** [comment]
@qiankehan please use use the command "avocado list --vt-type libvirt usb_device --vt-machine-type q35"
with the machine type q35

**Comment by qiankehan** [comment]
@brucepatricklee 
It looks something wrong in python syntaxt:
 (20/21) type_specific.io-github-autotest-libvirt.usb_device.redirdev.spicevmc.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: ERROR: Key 'type' not found in super class attributes or in ('type_name', 'type_name', 'device_tag', 'xml', 'virsh', 'xmltreefile', 'validates', 'get_type', 'set_type', 'del_type') (21.80 s)
 (21/21) type_specific.io-github-autotest-libvirt.usb_device.redirdev.tcp.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: ERROR: Key 'type' not found in super class attributes or in ('type_name', 'type_name', 'device_tag', 'xml', 'virsh', 'xmltreefile', 'validates', 'get_type', 'set_type', 'del_type') (23.04 s)
RESULTS    : PASS 19 | ERROR 2 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1635.53 s

**Comment by brucepatricklee** [comment]
@qiankehan as I mentioned in the previous comments, this commit should be merged with commit https://github.com/avocado-framework/avocado-vt/pull/2076, you need to patch this commit before run test case in your own machine

**Comment by qiankehan** [comment]
Works for me. Please remember to mention the required avocado-vt version for running the cases @brucepatricklee

File: libvirt/tests/src/usb_device.py [hunk 1]
```diff
@@ -109,37 +117,41 @@ def usb_disk_check(session, src_guest=None):
...
-
-    def usb_device_check(session, src_host=None):
+        # check whether the guest got the right usb device
+        output = output.strip().splitlines()
+        for guest in src_guest['product']:
+            pattern = "ID_MODEL_ID={}".format(guest['product_id'].lstrip("0x"))
+            for line in output:
+                if pattern in line:
+                    return
+        test.fail("guest got the wrong usb device")
```
Comment (kylazhang) [inline libvirt/tests/src/usb_device.py]: And for all failure better give exactly which pattern failed on which output for better debugging

File: libvirt/tests/src/usb_device.py [hunk 2]
```diff
@@ -214,95 +232,132 @@ def check_controller_alias(alias):
...
+        # assemble the xml of usb redirdev device
+        if device_name == "redirdev":
+            for i, addr in enumerate(src_host['product']):
+                device_xml = vmxml.get_device_class(device_name)()
+                device_xml.type = device_type
+                device_xml.bus = "usb"
+                if device_type == "tcp":
+                    source_xml = device_xml.Source()
+                    source_xml.mode = "connect"
+                    source_xml.host = "localhost"
```
Comment (kylazhang) [inline libvirt/tests/src/usb_device.py]: For the fixed value, we'd better setting in cfg

File: libvirt/tests/src/usb_device.py [hunk 3]
```diff
@@ -214,95 +232,132 @@ def check_controller_alias(alias):
...
-        if passthrough and hotplug:
-            for hostdev in hostdev_list:
-                virsh.attach_device(vm_name, hostdev.xml, flagstr="--live",
+        if usb_alias or redirdev_alias:
+            check_alias(device_alias)
+
+        # install package usbutils in guest
+        for pkg in pkgs_guest.split(','):
+            if not utils_package.package_install(pkg, session):
+                test.fail("package usbutils installation fails")
```
Comment (kylazhang) [inline libvirt/tests/src/usb_device.py]: As above

**Comment by brucepatricklee** [comment]
(p3-ve) [root@jil_rhel8 code]# avocado run --vt-type libvirt usb_device  --vt-machine-type q35 
JOB ID     : 7ffe1372c3a7434d3ec24f02bb55fc3168b0c463
JOB LOG    : /root/avocado/job-results/job-2019-06-04T11.05-7ffe137/job.log
 (01/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (93.40 s)
 (02/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (96.62 s)
 (03/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.nec_xhci_hub.pcie-root.dmi-to-pci-bridge: PASS (97.52 s)
 (04/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (97.89 s)
 (05/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (96.54 s)
 (06/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb3_0.qemu_xhci.pcie-root.dmi-to-pci-bridge: PASS (95.37 s)
 (07/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (99.87 s)
 (08/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (99.25 s)
 (09/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.vid_pid.usb2_0.companion.pcie-root.dmi-to-pci-bridge: PASS (97.71 s)
 (10/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (99.66 s)
 (11/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (89.99 s)
 (12/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.nec_xhci_hub.pcie-root.dmi-to-pci-bridge: PASS (93.08 s)
 (13/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (88.25 s)
 (14/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (87.18 s)
 (15/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb3_0.qemu_xhci.pcie-root.dmi-to-pci-bridge: PASS (91.55 s)
 (16/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (95.18 s)
 (17/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pcie-root.pcie-root-port.pcie-switch-upstream-port: PASS (105.88 s)
 (18/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb2_0.companion.pcie-root.dmi-to-pci-bridge: PASS (97.14 s)
 (19/21) type_specific.io-github-autotest-libvirt.usb_device.passthrough.bus_dev.usb_all.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (100.75 s)
 (20/21) type_specific.io-github-autotest-libvirt.usb_device.redirdev.spicevmc.usb2_0.companion.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (25.70 s)
 (21/21) type_specific.io-github-autotest-libvirt.usb_device.redirdev.tcp.usb2_0.companion.pcie-root.pcie-root-port.pcie-to-pci-bridge: PASS (165.13 s)
RESULTS    : PASS 21 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 2017.15 s
(p3-ve) [root@jil_rhel8 code]#

**Pull request merged**